### DISCUSSION TITLE
Security dependencies precedence order is not preserved

### DISCUSSION THREAD

**Discussion by nadobando** [fastapi/fastapi#8735] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import json

from fastapi import FastAPI, APIRouter, Security, security

token_scheme = security.OAuth2PasswordBearer(tokenUrl="/token", scopes={'me': ""})

app = FastAPI(dependencies=[Security(token_scheme)])

sub_router = APIRouter(prefix="/sub_router")


@app.get("/")
async def root():
    return {"message": "Hello World"}


@sub_router.get("/", dependencies=[Security(token_scheme, scopes=["me"])])
async def say_hello(name: str):
    return {"message": f"Hello {name}"}


app.include_router(sub_router)

json.dumps(app.openapi(), indent=2)
```

### Description

I've added a global security dependency to the app `token_scheme` then created a sub router `sub_router` which includes an endpoint which need a specific `oauth2` scope `me`.

when I get the `openapi` spec the endpoint `/sub_router/` does not have the relevant oauth2 scope `me`

```json
{
  "openapi": "3.0.2",
  "info": {
    "title": "FastAPI",
    "version": "0.1.0"
  },
  "paths": {
    "/": {
      "get": {
        "summary": "Root",
        "operationId": "root__get",
        "responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {}
              }
            }
          }
        },
        "security": [
          {
            "OAuth2PasswordBearer": []
          }
        ]
      }
    },
    "/sub_router/": {
      "get": {
        "summary": "Say Hello",
        "operationId": "say_hello_sub_router__get",
        "parameters": [
          {
            "required": true,
            "schema": {
              "title": "Name",
              "type": "string"
            },
            "name": "name",
            "in": "query"
          }
        ],
        "responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {}
              }
            }
          }
        },
        "security": [
          {
            "OAuth2PasswordBearer": []
          }
        ]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "OAuth2PasswordBearer": {
        "type": "oauth2",
        "flows": {
          "password": {
            "scopes": {
              "me": ""
            },
            "tokenUrl": "/token"
          }
        }
      }
    }
  }
}
```

what I expect to see

```json
"/sub_router/": {
      "get": {
        "summary": "Say Hello",
        "operationId": "say_hello_sub_router__get",
        "parameters": [
          {
            "required": true,
            "schema": {
              "title": "Name",
              "type": "string"
            },
            "name": "name",
            "in": "query"
          }
        ],
        "responses": {
          "200": {
            "description": "Successful Response",
            "content": {
              "application/json": {
                "schema": {}
              }
            }
          }
        },
        "security": [
          {
            "OAuth2PasswordBearer": ["me"]
          }
        ]
      }
    }
```

I've played a little bit with the `include_router` function and changed a little bit the order of `current_dependencies` and got to preserve the latest added security scope (I can create this PR). 
but maybe this should be solved in a different way since it may be useful to merge security scopes instead of overriding them and specify all scopes again in each endpoint dependencies

1. endpoint security dependencies
2. sub_router security dependencies
3. router security dependencies
4. app security dependencies

### Operating System

macOS

### FastAPI Version

0.71.0

### Python Version

3.9.5

**Comment by fpoirotte** [comment]
Hello,

I just stumbled upon a similar issue as well.
In my case the scopes specified in the endpoint security dependency got overwritten by those defined in the router (through a security dependency during the call to `include_router`).
I expected the scopes to be merged together.

Best regards,
François

**Reply by nadobando** [reply]
I guess it’s an option but not the one I prefer.
I would like this to be inverted, 
think about a package which defines a new scope and has this file `package/routes.py` and defines a sub-router there.
I would like to include the new scope as a sub-router dependency, then in the `app/main router` you can just include the new router and this will be merged as expected to the whole app this approach makes the package more autonomous, WDYT?

**Comment by YuriiMotov** [comment, marked as answer]
That was fixed at some point.
Now it's:

```
    "/sub_router/": {
      "get": {
        "summary": "Say Hello",
        "operationId": "say_hello_sub_router__get",
        "security": [
          {
            "OAuth2PasswordBearer": []
          },
          {
            "OAuth2PasswordBearer": [
              "me"
            ]
          }
        ],
```
